core/integration/module_test.go

```diff
@@ -1189,6 +1189,113 @@ func (m *Foo) MyFunction() X {
 	require.JSONEq(t, `{"foo":{"myFunction":{"message":"foo"}}}`, out)
 }
 
+func TestModuleGoReturnStruct(t *testing.T) {
+	t.Parallel()
+
+	c, ctx := connect(t)
+
+	modGen := c.Container().From(golangImage).
+		WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
+		WithWorkdir("/work").
+		With(daggerExec("mod", "init", "--name=foo", "--sdk=go")).
+		WithNewFile("main.go", dagger.ContainerWithNewFileOpts{
+			Contents: `package main
+
+type Foo struct {}
+
+type X struct {
+	Message string ` + "`json:\"message\"`" + `
+	To string ` + "`json:\"recipient\"`" + `
+	From string
+}
+
+func (m *Foo) MyFunction() X {
+	return X{Message: "foo", To: "user", From: "admin"}
+}
+`,
+		})
+
+	logGen(ctx, t, modGen.Directory("."))
+
+	out, err := modGen.With(daggerQuery(`{foo{myFunction{message, recipient, from}}}`)).Stdout(ctx)
+	require.NoError(t, err)
+	require.JSONEq(t, `{"foo":{"myFunction":{"message":"foo", "recipient":"user", "from":"admin"}}}`, out)
+}
+
+func TestModuleGoReturnNestedStruct(t *testing.T) {
+	t.Parallel()
+
+	c, ctx := connect(t)
+
+	modGen := c.Container().From(golangImage).
+		WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
+		WithWorkdir("/work").
+		With(daggerExec("mod", "init", "--name=playground", "--sdk=go")).
+		WithNewFile("main.go", dagger.ContainerWithNewFileOpts{
+			Contents: `package main
+
+type Playground struct{}
+
+type Foo struct {
+	MsgContainer Bar
```

Comment:
I think it would be good to have test coverage for a case that's essentially a complex combination of all 3 new test cases you added:
1. Mix of `json:` usage and not
2. Use of complex/nested types
3. Use of core ID able types like Container

The coverage of those individually is great but it's the interactions between all the different cases that gets especially hard to reason through just reading the code, so integ tests are especially valuable for there.

I don't have a strong opinion on whether that means combining everything into one test case or keeping these 3 "focused" ones and adding an additional one that tests every in use at once; whichever feels best.

Reply:
Done, added a new test case - no changes required to make it work :tada: